File: generate_hex_games/hex_generator.py
import random
import csv

BOARD_DIM = 5

# Neighbor offsets for a hexagonal grid
neighbors = [-(BOARD_DIM + 2) + 1, -(BOARD_DIM + 2), -1, 1, (BOARD_DIM + 2), (BOARD_DIM + 2) - 1]

class HexGame:
    def __init__(self):
        self.board = [0] * ((BOARD_DIM + 2) * (BOARD_DIM + 2))
        self.open_positions = []
        self.moves = []  # Track the moves to allow for rewinding
        self.init_board()

    def init_board(self):
        """Initialize the board and open positions."""
        self.open_positions = []
        for i in range(1, BOARD_DIM + 1):
            for j in range(1, BOARD_DIM + 1):
                self.open_positions.append(i * (BOARD_DIM + 2) + j)

    def place_piece_randomly(self, player):
        """Place a piece randomly on the board."""
        if not self.open_positions:
            return None

        random_empty_position = random.choice(self.open_positions)
        self.board[random_empty_position] = player
        self.moves.append((random_empty_position, player))  # Track the move
        self.open_positions.remove(random_empty_position)
        return random_empty_position

    def remove_last_move(self):
        """Remove the last move to rewind the game."""
        if not self.moves:
            return
        last_move, player = self.moves.pop()
        self.board[last_move] = 0  # Remove the last piece
        self.open_positions.append(last_move)  # Restore the open position

    def rewind_moves(self, num_moves):
        """Rewind the game by a specific number of moves."""
        for _ in range(num_moves):
            if self.moves:
                self.remove_last_move()

    def dfs(self, player, position, visited):
        """DFS to explore connectivity for the given player."""
        visited[position] = True
        row = position // (BOARD_DIM + 2)
        col = position % (BOARD_DIM + 2)

        # Player 1 (O) wins by connecting top to bottom
        if player == 1 and row == BOARD_DIM:
            return True

        # Player -1 (X) wins by connecting left to right
        if player == -1 and col == BOARD_DIM:
            return True

        # Explore neighbors
        for neighbor_offset in neighbors:
            neighbor = position + neighbor_offset
            if 0 <= neighbor < (BOARD_DIM + 2) * (BOARD_DIM + 2) and not visited[neighbor]:
                if self.board[neighbor] == player:
                    if self.dfs(player, neighbor, visited):
                        return True
        return False

    def check_guaranteed_win(self, player):
        """Check if the player is in a guaranteed winning position."""
        visited = [False] * ((BOARD_DIM + 2) * (BOARD_DIM + 2))

        if player == 1:
            # Player 1 must connect top to bottom
            for col in range(1, BOARD_DIM + 1):
                start_pos = (BOARD_DIM + 2) + col  # Top row
                if self.board[start_pos] == player and not visited[start_pos]:
                    if self.dfs(player, start_pos, visited):
                        return True
        else:
            # Player -1 must connect left to right
            for row in range(1, BOARD_DIM + 1):
                start_pos = row * (BOARD_DIM + 2) + 1  # Leftmost column
                if self.board[start_pos] == player and not visited[start_pos]:
                    if self.dfs(player, start_pos, visited):
                        return True
        return False

    def full_board(self):
        """Check if the board is full."""
        return len(self.open_positions) == 0

    def get_board_state(self):
        """Return the board state in a column-major order for CSV saving."""
        board_state = []
        for col in range(1, BOARD_DIM + 1):
            for row in range(1, BOARD_DIM + 1):
                board_state.append(self.board[row * (BOARD_DIM + 2) + col])
        return board_state

def save_game_to_csv(game, winner, csv_writer):
    """Save the board state and the winner to a CSV file."""
    board_state = game.get_board_state()
    csv_writer.writerow(board_state + [winner])

def write_csv_headers(csv_writer):
    """Write the CSV headers in the format cell{col}_{row} and winner."""
    headers = []
    for col in range(BOARD_DIM):
        for row in range(BOARD_DIM):
            headers.append(f"cell{col}_{row}")
    headers.append("winner")
    csv_writer.writerow(headers)

def simulate_hex_game(game):
    """Simulate a Hex game until one player wins."""
    current_player = 1  # Player 1 starts (O)

    while not game.full_board():
        game.place_piece_randomly(current_player)
        if game.check_guaranteed_win(current_player):
            return current_player  # Return the winner as soon as they have a guaranteed victory

        current_player = -current_player  # Switch player

    return 0  # Shouldn't happen in a complete game

# Simulate and save games to CSV
if __name__ == "__main__":
    num_games_per_category = 1000 // 6  # 1000 total games, equally divided across categories
    player_1_wins = 0
    player_neg1_wins = 0

    with open('hex_winning_positions.csv', mode='w', newline='') as file:
        csv_writer = csv.writer(file)
        write_csv_headers(csv_writer)

        for rewind_moves in range(0, 6):  # 0 = fully connected, 1 to 5 moves away
            for game_num in range(num_games_per_category):
                game = HexGame()

                winner = simulate_hex_game(game)  # Simulate the full game
                if rewind_moves > 0:
                    # Rewind 1 to 5 moves to create near-winning states
                    game.rewind_moves(rewind_moves)

                # Save the game result to the CSV
                save_game_to_csv(game, winner, csv_writer)

                if winner == 1:
                    player_1_wins += 1
                elif winner == -1:
                    player_neg1_wins += 1

                if game_num % 100 == 0:
                    print(f"Rewind {rewind_moves} - Game {game_num} completed.")

    print(f"Player 1 wins: {player_1_wins}")
    print(f"Player -1 wins: {player_neg1_wins}")

File: generate_hex_games/hex_generator.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifndef BOARD_DIM
#define BOARD_DIM 5
#endif

// Neighboring cell offsets for hexagonal grid.
int neighbors[] = {-(BOARD_DIM + 2) + 1, -(BOARD_DIM + 2), -1, 1, (BOARD_DIM + 2), (BOARD_DIM + 2) - 1};

struct hex_game {
    int board[(BOARD_DIM + 2) * (BOARD_DIM + 2)];
    int open_positions[BOARD_DIM * BOARD_DIM];
    int number_of_open_positions;
};

// Initialize the hex game with empty cells and open positions.
void hg_init(struct hex_game *hg) {
    for (int i = 0; i < BOARD_DIM + 2; ++i) {
        for (int j = 0; j < BOARD_DIM + 2; ++j) {
            hg->board[i * (BOARD_DIM + 2) + j] = 0;
            if (i > 0 && i < BOARD_DIM + 1 && j > 0 && j < BOARD_DIM + 1) {
                hg->open_positions[(i - 1) * BOARD_DIM + j - 1] = i * (BOARD_DIM + 2) + j;
            }
        }
    }
    hg->number_of_open_positions = BOARD_DIM * BOARD_DIM;
}

// Recursive DFS function to check connections for a player.
int hg_dfs(struct hex_game *hg, int player, int position, int *visited) {
    visited[position] = 1;  // Mark this position as visited.
    int row = position / (BOARD_DIM + 2);
    int col = position % (BOARD_DIM + 2);

    // Check if player 1 reached the right edge or player -1 reached the bottom edge.
    if ((player == 1 && col == BOARD_DIM) || (player == -1 && row == BOARD_DIM)) {
        return 1;
    }

    // Explore all neighbors.
    for (int i = 0; i < 6; ++i) {
        int neighbor = position + neighbors[i];
        if (neighbor >= 0 && neighbor < (BOARD_DIM + 2) * (BOARD_DIM + 2) &&
            hg->board[neighbor] == player && !visited[neighbor]) {
            if (hg_dfs(hg, player, neighbor, visited)) {
                return 1;
            }
        }
    }

    return 0;
}

// Function to check if a player has won.
int hg_winner(struct hex_game *hg, int player) {
    int visited[(BOARD_DIM + 2) * (BOARD_DIM + 2)];
    memset(visited, 0, sizeof(visited));  // Reset visited array.

    if (player == 1) {
        // Check if player 1 has connected left to right.
        for (int row = 1; row <= BOARD_DIM; ++row) {
            int start_pos = row * (BOARD_DIM + 2) + 1;  // Starting from the left edge.
            if (hg->board[start_pos] == 1 && hg_dfs(hg, 1, start_pos, visited)) {
                return 1;  // Player 1 wins.
            }
        }
    } else {
        // Check if player -1 has connected top to bottom.
        for (int col = 1; col <= BOARD_DIM; ++col) {
            int start_pos = (BOARD_DIM + 2) + col;  // Starting from the top edge.
            if (hg->board[start_pos] == -1 && hg_dfs(hg, -1, start_pos, visited)) {
                return -1;  // Player -1 wins.
            }
        }
    }

    return 0;
}

// Place a piece randomly on the board.
int hg_place_piece_randomly(struct hex_game *hg, int player) {
    int random_empty_position_index = rand() % hg->number_of_open_positions;
    int empty_position = hg->open_positions[random_empty_position_index];
    hg->board[empty_position] = player;
    hg->open_positions[random_empty_position_index] = hg->open_positions[hg->number_of_open_positions - 1];
    hg->number_of_open_positions--;
    return empty_position;
}

// Check if the board is full.
int hg_full_board(struct hex_game *hg) {
    return hg->number_of_open_positions == 0;
}

// Save the game state and winner to a CSV file.
void save_game_to_csv(struct hex_game *hg, int winner, FILE *file) {
    for (int col = 1; col <= BOARD_DIM; ++col) {
        for (int row = 1; row <= BOARD_DIM; ++row) {
            int value = hg->board[row * (BOARD_DIM + 2) + col];
            fprintf(file, "%d,", value);
        }
    }
    fprintf(file, "%d\n", winner);
}

// Write the CSV headers.
void write_csv_headers(FILE *file) {
    for (int col = 0; col < BOARD_DIM; ++col) {
        for (int row = 0; row < BOARD_DIM; ++row) {
            fprintf(file, "cell%d_%d,", col, row);
        }
    }
    fprintf(file, "winner\n");
}

int main() {
    struct hex_game hg;
    FILE *csv_file = fopen("hex_games_11.csv", "w");
    if (csv_file == NULL) {
        printf("Error opening file!\n");
        return 1;
    }
    write_csv_headers(csv_file);

    for (int game = 0; game < 1000; ++game) {
        hg_init(&hg);
        int player = 1;  // Player 1 starts (O).
        int winner = 0;

        while (!hg_full_board(&hg)) {
            int position = hg_place_piece_randomly(&hg, player);
            winner = hg_winner(&hg, player);
            if (winner != 0) {
                break;
            }
            player = -player;  // Switch player.
        }

        save_game_to_csv(&hg, winner, csv_file);

        if (game % 100 == 0) {
            printf("Game %d completed.\n", game);
        }
    }

    fclose(csv_file);
    return 0;
}
